x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/EsqlResponseListener.java

```diff
@@ -26,7 +27,60 @@
 import static org.elasticsearch.xpack.esql.formatter.TextFormat.CSV;
 import static org.elasticsearch.xpack.esql.formatter.TextFormat.URL_PARAM_DELIMITER;
 
+/**
+ * Listens for a single {@link EsqlQueryResponse}, builds a corresponding {@link RestResponse} and sends it.
+ */
 public class EsqlResponseListener extends RestResponseListener<EsqlQueryResponse> {
+    /**
+     * A simple, thread-safe stop watch for timing a single action.
+     * Allows to stop the time for building a response and to e.g. log it at a later point.
+     */
+    private static class ThreadSafeStopWatch {
+        /**
+         * Start time of the watch
+         */
+        private long startTimeNS;
+
+        /**
+         * End time of the watch
+         */
+        private long endTimeNS;
+
+        /**
+         * Is the stop watch currently running?
+         */
+        private boolean running;
+
+        /**
+         * Start the stop watch. Will do nothing if it is already running.
+         */
+        public void start() {
+            synchronized (this) {
+                if (running) {
+                    return;
+                } else {
+                    startTimeNS = System.nanoTime();
+                    running = true;
+                }
+            }
+        }
+
+        /**
+         * Stop the stop watch (or do nothing if it is not running) and return the elapsed time since starting.
+         * @return the elapsed time since starting the watch
+         */
+        public TimeValue stop() {
+            synchronized (this) {
```

Comment:
No need for synchronization here as the stop will be called by a single thread.

Reply:
I think removing the start method and starting the stopwatch on construction is fine with respect to thread-safety.

However, I don't think the stop method can ever be truly thread safe without some kind of synchronization. While this may never happen in practice, two threads _could_ try to stop the watch at roughly the same time, and the `long` value `endTimeNS` might end up being inconsistent; or it might be read before being initialized at all; or, stopping the watch in one thread might not propagate to another thread (as `running` and `endTimeNS` are not `volatile`).

Because of this, I'd like to keep the synchronization on `stop` to rule out any doubt. The synchronization overhead is likely going to be minimal, because I _think_ that any given instance of `EsqlResponseListener` is only used for 1 request (some request params are used in `EsqlResponseListener`'s constructor), so there shouldn't be contention on the lock.

Reply:
if stop can only ever be called by a single thread, and just once, then we can of course drop the synchronization. Otherwise, we need it, right.